internal/mode/static/telemetry/collector_test.go

```diff
@@ -0,0 +1,346 @@
+						{Namespace: "test", Name: "hr-2"}: {},
+						{Namespace: "test", Name: "hr-3"}: {},
+					},
+					ReferencedSecrets: map[types.NamespacedName]*graph.Secret{
+						client.ObjectKeyFromObject(secret1): {
+							Source: secret1,
+						},
+						client.ObjectKeyFromObject(secret2): {
+							Source: secret2,
+						},
+						client.ObjectKeyFromObject(nilsecret): nil,
+					},
+					ReferencedServices: map[types.NamespacedName]struct{}{
+						client.ObjectKeyFromObject(svc1):   {},
+						client.ObjectKeyFromObject(svc2):   {},
+						client.ObjectKeyFromObject(nilsvc): {},
+					},
+				}
+
+				config := &dataplane.Configuration{
+					Upstreams: []dataplane.Upstream{
+						{
+							Name:     "upstream1",
+							ErrorMsg: "",
+							Endpoints: []resolver.Endpoint{
+								{
+									Address: "endpoint1",
+									Port:    80,
+								}, {
+									Address: "endpoint2",
+									Port:    80,
+								}, {
+									Address: "endpoint3",
+									Port:    80,
+								},
+							},
+						},
+						{
+							Name:     "upstream2",
+							ErrorMsg: "",
+							Endpoints: []resolver.Endpoint{
+								{
+									Address: "endpoint1",
+									Port:    80,
+								},
+							},
+						},
+						{
+							Name:     "upstream3",
+							ErrorMsg: "there is an error here",
+							Endpoints: []resolver.Endpoint{
+								{
+									Address: "endpoint1",
+									Port:    80,
+								}, {
+									Address: "endpoint2",
+									Port:    80,
+								}, {
+									Address: "endpoint3",
+									Port:    80,
+								},
+							},
+						},
+					},
+				}
+
+				fakeGraphGetter.GetLatestGraphReturns(graph)
+				fakeConfigurationGetter.GetLatestConfigurationReturns(config)
+
+				expData.NodeCount = 3
+				expData.NGFResourceCounts = telemetry.NGFResourceCounts{
+					Gateways:       3,
+					GatewayClasses: 3,
+					HTTPRoutes:     3,
+					Secrets:        3,
+					Services:       3,
+					Endpoints:      4,
+				}
+
+				data, err := dataCollector.Collect(ctx)
+
+				Expect(err).To(BeNil())
+				Expect(expData).To(Equal(data))
+			})
+		})
+	})
+
+	Describe("node count collector", func() {
+		When("collecting node count data", func() {
+			It("collects correct data for no nodes", func() {
+				k8sClientReader.ListCalls(func(ctx context.Context, list client.ObjectList, option ...client.ListOption) error {
+					Expect(option).To(BeEmpty())
+
+					switch typedList := list.(type) {
+					case *v1.NodeList:
+						typedList.Items = []v1.Node{}
+					default:
+						Fail(fmt.Sprintf("unknown type: %T", typedList))
+					}
+					return nil
+				})
+
+				data, err := dataCollector.Collect(ctx)
+
+				Expect(err).To(BeNil())
+				Expect(expData).To(Equal(data))
+			})
+
+			It("collects correct data for one node", func() {
+				node := v1.Node{
+					ObjectMeta: metav1.ObjectMeta{Name: "node1"},
+				}
+
+				k8sClientReader.ListCalls(func(ctx context.Context, list client.ObjectList, option ...client.ListOption) error {
```

Comment:
If you want to reduce code duplication, you could create a helper function that returns a `ListCalls` function with the given nodes:

```
func createListCallsFunc(nodes []v1.Node) func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
    return func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
         // the function body
    }
}
```

Reply:
i love this!